File: backend/app/routers/billing.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.user import User
from app.schemas.billing import InvoiceSummary, LogOut, RechargeRequest, UsageStats
from app.services import billing_service

router = APIRouter(prefix="/api/v1/billing", tags=["billing"])


@router.get("/usage", response_model=UsageStats)
async def usage(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await billing_service.get_usage_stats(db, user)


@router.get("/logs")
async def logs(
    page: int = 1,
    page_size: int = 20,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    from sqlalchemy import func, select
    from app.models.request_log import RequestLog
    total = (await db.execute(
        select(func.count(RequestLog.id)).where(RequestLog.user_id == user.id)
    )).scalar() or 0
    items = await billing_service.get_logs(db, user, page, page_size)
    return {"items": items, "total": total, "page": page, "page_size": page_size}


@router.get("/balance")
async def balance(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    from sqlalchemy import func, select
    from app.models.payment import Payment
    from app.models.request_log import RequestLog
    await db.refresh(user)
    total_recharged = (await db.execute(
        select(func.coalesce(func.sum(Payment.amount), 0)).where(Payment.user_id == user.id)
    )).scalar() or 0
    total_consumed = (await db.execute(
        select(func.coalesce(func.sum(RequestLog.cost), 0)).where(RequestLog.user_id == user.id)
    )).scalar() or 0
    return {
        "balance": float(user.balance),
        "total_recharged": float(total_recharged),
        "total_consumed": float(total_consumed),
    }


@router.post("/recharge")
async def recharge(
    req: RechargeRequest,
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    try:
        return await billing_service.recharge(db, user, req.code)
    except ValueError as e:
        raise HTTPException(status_code=400, detail={"error": str(e)})


@router.get("/invoices", response_model=list[InvoiceSummary])
async def invoices(
    period: str = "daily",
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    return await billing_service.get_invoices(db, user, period)

File: frontend/src/api/billing.ts
import client from './client';

export interface UsageStats {
  total_requests: number;
  total_tokens: number;
  total_cost: number;
  prompt_tokens: number;
  completion_tokens: number;
  total_prompt_tokens?: number;
  total_completion_tokens?: number;
  reasoning_tokens: number;
  daily_stats?: any[];
}

export interface RequestLog {
  id: string;
  model: string;
  prompt_tokens: number;
  completion_tokens: number;
  reasoning_tokens: number;
  cost: number;
  latency_ms: number;
  status: string;
  error_message: string | null;
  created_at: string;
}

export interface LogsResponse {
  items: RequestLog[];
  total: number;
  page: number;
  page_size: number;
}

export interface BalanceInfo {
  balance: number;
  total_recharged?: number;
  total_consumed?: number;
}

export interface RechargeRequest {
  code: string;
}

export interface RechargeResponse {
  amount: number;
  new_balance?: number;
  balance?: number;
}

export interface Invoice {
  period: string;
  total_requests: number;
  total_tokens: number;
  total_cost: number;
}

export async function getUsage(startDate?: string, endDate?: string): Promise<UsageStats> {
  const params: Record<string, string> = {};
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const res = await client.get<UsageStats>('/billing/usage', { params });
  return res.data;
}

export async function getLogs(
  page: number = 1,
  pageSize: number = 20,
  startDate?: string,
  endDate?: string
): Promise<LogsResponse | RequestLog[]> {
  const params: Record<string, string | number> = { page, page_size: pageSize };
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const res = await client.get('/billing/logs', { params });
  return res.data;
}

export async function getBalance(): Promise<BalanceInfo> {
  const res = await client.get<BalanceInfo>('/billing/balance');
  return res.data;
}

export async function recharge(data: RechargeRequest): Promise<RechargeResponse> {
  const res = await client.post<RechargeResponse>('/billing/recharge', data);
  return res.data;
}

export async function getInvoices(
  period: 'daily' | 'weekly' | 'monthly' = 'daily',
  startDate?: string,
  endDate?: string
): Promise<Invoice[]> {
  const params: Record<string, string> = { period };
  if (startDate) params.start_date = startDate;
  if (endDate) params.end_date = endDate;
  const res = await client.get<Invoice[]>('/billing/invoices', { params });
  return res.data;
}
